File: scripts/analyze_asset_requirements.py
#!/usr/bin/env python3
"""
素材资源需求分析脚本

分析 data/era_assets/ 目录，生成资源缺口报告
对应 docs/plans/2026-05-04_asset-resource-detailed-requirements.md

Usage:
    python3 scripts/analyze_asset_requirements.py
    python3 scripts/analyze_asset_requirements.py --json
    python3 scripts/analyze_asset_requirements.py --era ancient_eastern_wuxia
"""

import os
import sys
import json
import argparse
from pathlib import Path
from datetime import datetime
from typing import Dict, List, Optional, Any

# Paths
ERA_ASSETS_DIR = Path(__file__).parent.parent / "data" / "era_assets"
OUTPUT_DIR = Path(__file__).parent.parent / ".asset-reports"

# Era Name Mapping
ERA_NAME_MAP = {
    'ancient_eastern_conspiracy': '古代·权谋',
    'ancient_eastern_myth': '古代·神话',
    'ancient_eastern_wuxia': '古代·武侠',
    'ancient_eastern_xianxia': '古代·修仙',
    'ancient_eastern_zhiguai': '古代·志怪',
    'ancient_prehistoric': '古代·史前',
    'ancient_western': '古代·西方',
    'ancient_western_celtic': '古代·凯尔特',
    'ancient_western_greek': '古代·古希腊',
    'ancient_western_medieval': '古代·中世纪',
    'ancient_western_roman': '古代·古罗马',
    'ancient_western_viking': '古代·维京',
    'apocalyptic_nuclear': '末日·核战',
    'apocalyptic_plague': '末日·瘟疫',
    'apocalyptic_post_war': '末日·战后',
    'contemporary_apocalypse': '现代·末日',
    'contemporary_campus': '现代·校园',
    'contemporary_extreme_cold': '现代·极寒',
    'contemporary_hippie': '现代·嬉皮士',
    'contemporary_noir': '现代·黑色电影',
    'contemporary_nuclear_winter': '现代·核冬天',
    'contemporary_post_apocalyptic': '现代·末日废土',
    'contemporary_rural': '现代·乡村',
    'contemporary_urban': '现代·都市',
    'contemporary_western': '现代·西部',
    'contemporary_zombie': '现代·生化危机',
    'cyberpunk': '赛博朋克',
    'fantasy_renaissance': '幻想·文艺复兴',
    'far-future': '远未来',
    'far-future_cyborg': '未来·义体化',
    'far-future_space_opera': '未来·太空歌剧',
    'far-future_virtual_reality': '未来·虚拟现实',
    'future_cyberpunk': '未来·赛博',
    'future_near_term': '未来·近未来',
    'future_space_colony': '未来·太空殖民',
    'modern_early_industrial': '近代·早期工业',
    'modern_early_republic': '近代·早期共和',
    'modern_eastern_meiji_taisho': '近代·明治大正',
    'modern_eastern_republic': '近代·民国',
    'modern_steampunk': '近代·蒸汽朋克',
    'modern_wartime': '近代·战时',
    'modern_western_jazz_age': '近代·爵士时代',
    'modern_western_postwar': '近代·战后',
    'modern_western_victorian': '近代·维多利亚',
    'near-future_cyberpunk': '近未来·赛博朋克',
    'near-future_dystopia': '近未来·反乌托邦',
    'near-future_space_colonization': '近未来·太空殖民',
    'post_human': '后人类',
    'post_human_ai': '后人类·AI觉醒',
    'post_human_transhuman': '后人类·基因改造',
    'post_human_virtual': '后人类·虚拟',
    'prehistoric_dinosaur': '原生·恐龙时代',
    'prehistoric_ice_age': '原生·冰河时代',
    'prehistoric_primeval': '原生·蛮荒时代',
    'renaissance': '文艺复兴',
    'school_sim': '校园模拟',
    'space_scifi': '太空科幻',
}

def get_era_name(era_id: str) -> str:
    return ERA_NAME_MAP.get(era_id, era_id.replace('_', ' ').replace('-', ' '))

def load_manifest(era_dir: Path) -> Optional[Dict]:
    manifest_path = era_dir / "manifest.json"
    if not manifest_path.exists():
        return None
    try:
        with open(manifest_path, 'r', encoding='utf-8') as f:
            return json.load(f)
    except Exception as e:
        print(f"[WARN] Failed to load manifest {manifest_path}: {e}")
        return None

def get_manifest_format(manifest: Optional[Dict]) -> str:
    if manifest is None:
        return 'none'
    if 'era_id' in manifest:
        return 'new'
    if 'id' in manifest:
        return 'old'
    return 'unknown'

def count_images(manifest: Optional[Dict]) -> int:
    if manifest is None:
        return 0
    images = manifest.get('images', [])
    return len(images)

def has_bgm(manifest: Optional[Dict]) -> bool:
    if manifest is None:
        return False
    bgm = manifest.get('bgm')
    if bgm is None:
        return False
    if isinstance(bgm, list):
        return len(bgm) > 0
    return bool(bgm)

def get_status(manifest: Optional[Dict], fmt: str) -> str:
    if manifest is None:
        return 'missing'
    if 'status' in manifest:
        return manifest['status']
    if fmt == 'new':
        img_count = count_images(manifest)
        return 'complete' if img_count >= 6 else ('partial' if img_count > 0 else 'pending')
    return 'unknown'

def analyze_eras() -> List[Dict]:
    results = []
    
    if not ERA_ASSETS_DIR.exists():
        print(f"[ERROR] Era assets directory not found: {ERA_ASSETS_DIR}")
        return results
    
    for entry in sorted(ERA_ASSETS_DIR.iterdir()):
        if not entry.is_dir() or entry.name == 'index.html':
            continue
        
        manifest = load_manifest(entry)
        fmt = get_manifest_format(manifest)
        status = get_status(manifest, fmt)
        
        files = [f.name for f in entry.iterdir() 
                 if f.is_file() and not f.name.endswith('.json')]
        
        img_count = count_images(manifest)
        bgm_exists = has_bgm(manifest)
        is_complete = img_count >= 6 and bgm_exists
        needs_attention = status in ('missing', 'pending', 'partial')
        
        era_id = None
        era_name = None
        if manifest:
            era_id = manifest.get('era_id') or manifest.get('id')
            if 'era_name' in manifest:
                era_name = manifest.get('era_name')
        
        results.append({
            'name': entry.name,
            'id': era_id or entry.name,
            'era_name': era_name or get_era_name(entry.name),
            'has_manifest': fmt != 'none',
            'manifest_format': fmt,
            'status': status,
            'image_count': img_count,
            'has_bgm': bgm_exists,
            'files': files,
            'is_complete': is_complete,
            'needs_attention': needs_attention,
        })
    
    return results

def generate_summary(eras: List[Dict]) -> Dict:
    complete = sum(1 for e in eras if e['status'] == 'complete')
    partial = sum(1 for e in eras if e['status'] == 'partial')
    pending = sum(1 for e in eras if e['status'] == 'pending')
    missing = sum(1 for e in eras if e['status'] == 'missing')
    
    return {
        'era_scenes': {
            'total': len(eras),
            'complete': complete,
            'partial': partial,
            'pending': pending,
            'missing': missing,
        },
        'item_icons': {'total': 60, 'current': 14, 'gap': 46},
        'building_icons': {'total': 21, 'current': 4, 'gap': 17},
        'npc_avatars': {'total': 34},
        'skill_icons': {'total': 28, 'current': 1, 'gap': 27},
        'ui_icons': {'total': 37, 'current': 20, 'gap': 17},
        'bgm': {
            'scene': {'total': 36, 'current': 6, 'gap': 30},
            'era': {'total': 42, 'current': complete, 'gap': 42 - complete},
        },
        'sfx': {'total': 58, 'current': 15, 'gap': 43},
        'covers': {'total': 23},
    }

def print_report(eras: List[Dict], summary: Dict) -> None:
    print('╔══════════════════════════════════════════════════════════════════╗')
    print('║         素材资源需求分析报告                                      ║')
    print('║         Asset Resource Requirements Analysis                       ║')
    print('╚══════════════════════════════════════════════════════════════════╝')
    print()
    
    print('📊 时代素材总览 (Era Scene Assets Overview)')
    print('─' * 70)
    print(f"  总计子纪元: {len(eras)}")
    print(f"  ✅ 已完成 (complete):   {summary['era_scenes']['complete']}")
    print(f"  ⚠️  部分完成 (partial): {summary['era_scenes']['partial']}")
    print(f"  ⏳ 待生成 (pending):   {summary['era_scenes']['pending']}")
    print(f"  ❌ 缺失 (missing):     {summary['era_scenes']['missing']}")
    print()
    
    print('📋 各时代素材状态 (By Era)')
    print('─' * 70)
    
    for era in eras:
        status_icon = '✅' if era['is_complete'] else ('⚠️' if era['needs_attention'] else '❌')
        fmt_label = {'new': '[NEW]', 'old': '[OLD]', 'none': '[NO-MANIFEST]', 'unknown': '[?]'}
        format_str = fmt_label.get(era['manifest_format'], '[?]')
        
        img_info = f"{era['image_count']}/6 images"
        bgm_info = '🎵' if era['has_bgm'] else '❌'
        
        name_display = era.get('era_name', era['name'])
        print(f"  {status_icon} {format_str} {name_display:<20} {era['name']:<30} {img_info:<12} {bgm_info}")
    
    print()
    print('📦 其他资源需求 (Other Asset Requirements)')
    print('─' * 70)
    print(f"  物品图标: {summary['item_icons']['current']}/{summary['item_icons']['total']} (缺口: {summary['item_icons']['gap']})")
    print(f"  建筑图标: {summary['building_icons']['current']}/{summary['building_icons']['total']} (缺口: {summary['building_icons']['gap']})")
    print(f"  NPC头像: {summary['npc_avatars']['total']} (fallback)")
    print(f"  技能图标: {summary['skill_icons']['current']}/{summary['skill_icons']['total']} (缺口: {summary['skill_icons']['gap']})")
    print(f"  UI图标: {summary['ui_icons']['current']}/{summary['ui_icons']['total']} (缺口: {summary['ui_icons']['gap']})")
    print(f"  BGM: 场景{summary['bgm']['scene']['total']}首 + 时代{summary['bgm']['era']['total']}首")
    print(f"  SFX: {summary['sfx']['current']}/{summary['sfx']['total']} (缺口: {summary['sfx']['gap']})")
    print()
    
    # Priority items
    priority_eras = [e for e in eras if e['needs_attention']][:10]
    if priority_eras:
        print('🎯 优先补齐项目 (Priority Items)')
        print('─' * 70)
        for era in priority_eras:
            print(f"  • {era['name']} ({era['status']}) - {era.get('era_name', '')}")
        print()

def save_json_report(eras: List[Dict], summary: Dict) -> None:
    OUTPUT_DIR.mkdir(exist_ok=True)
    
    report = {
        'generated_at': datetime.now().isoformat(),
        'summary': summary,
        'eras': eras,
    }
    
    output_path = OUTPUT_DIR / f"asset-requirements-{datetime.now().strftime('%Y%m%d-%H%M%S')}.json"
    with open(output_path, 'w', encoding='utf-8') as f:
        json.dump(report, f, ensure_ascii=False, indent=2)
    print(f"[INFO] JSON report saved: {output_path}")
    
    latest_path = OUTPUT_DIR / "latest.json"
    with open(latest_path, 'w', encoding='utf-8') as f:
        json.dump(report, f, ensure_ascii=False, indent=2)
    print(f"[INFO] Latest report saved: {latest_path}")

def main():
    parser = argparse.ArgumentParser(description='素材资源需求分析')
    parser.add_argument('--json', action='store_true', help='Output JSON report')
    parser.add_argument('--era', type=str, help='Filter by era name')
    args = parser.parse_args()
    
    print('[INFO] Analyzing era assets...')
    eras = analyze_eras()
    
    if args.era:
        eras = [e for e in eras if args.era in e['name'] or args.era in e.get('era_name', '')]
        print(f'[INFO] Filtered to: {args.era}')
    
    summary = generate_summary(eras)
    
    if args.json:
        save_json_report(eras, summary)
    else:
        print_report(eras, summary)

if __name__ == '__main__':
    main()

File: scripts/analyze_asset_requirements.ts
#!/usr/bin/env npx ts-node
/**
 * 素材资源需求分析脚本
 * 
 * 分析 data/era_assets/ 目录，生成资源缺口报告
 * 对应 docs/plans/2026-05-04_asset-resource-detailed-requirements.md
 * 
 * Usage:
 *   npx ts-node scripts/analyze_asset_requirements.ts
 *   npx ts-node scripts/analyze_asset_requirements.ts --json
 *   npx ts-node scripts/analyze_asset_requirements.ts --era ancient_eastern_wuxia
 */

import * as fs from 'fs';
import * as path from 'path';
import { fileURLToPath } from 'url';
import * as yaml from 'js-yaml';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// ── Paths ──────────────────────────────────────────────────────────────────────

const ERA_ASSETS_DIR = path.resolve(__dirname, '../data/era_assets');
const OUTPUT_DIR = path.resolve(__dirname, '../.asset-reports');

// ── Types ─────────────────────────────────────────────────────────────────────

interface EraManifest {
  era_id?: string;
  era_name?: string;
  id?: string;
  status?: 'complete' | 'partial' | 'pending' | string;
  images?: (string | { id: string; cdn_url?: string; local_path?: string })[];
  bgm?: (string | { id: string; cdn_url?: string; local_path?: string })[];
  version?: string;
  updated_at?: string;
}

interface EraAnalysis {
  name: string;
  id: string;
  has_manifest: boolean;
  manifest_format: 'old' | 'new' | 'none';
  status: string;
  image_count: number;
  has_bgm: boolean;
  bgm_name?: string;
  files: string[];
  is_complete: boolean;
  needs_attention: boolean;
}

interface CategorySummary {
  total: number;
  complete: number;
  partial: number;
  pending: number;
  missing: number;
}

interface AssetRequirements {
  era_scenes: CategorySummary;
  item_icons: { total: number; current: number; gap: number };
  building_icons: { total: number; current: number; gap: number };
  npc_avatars: { total: number };
  skill_icons: { total: number; current: number; gap: number };
  ui_icons: { total: number; current: number; gap: number };
  bgm: { scene: CategorySummary; era: CategorySummary };
  sfx: { total: number; current: number; gap: number };
}

// ── Utility Functions ─────────────────────────────────────────────────────────

function getManifestFormat(manifestPath: string): 'old' | 'new' | 'none' {
  if (!fs.existsSync(manifestPath)) return 'none';
  
  try {
    const content = fs.readFileSync(manifestPath, 'utf-8');
    const manifest = JSON.parse(content) as EraManifest;
    
    if (manifest.era_id) return 'new';
    if (manifest.id) return 'old';
    return 'none';
  } catch {
    return 'none';
  }
}

function countImages(manifest: EraManifest): number {
  if (!manifest.images) return 0;
  return manifest.images.length;
}

function hasBgm(manifest: EraManifest | null): boolean {
  if (!manifest || !manifest.bgm) return false;
  if (Array.isArray(manifest.bgm)) return manifest.bgm.length > 0;
  return true;
}

function getBgmName(manifest: EraManifest): string | undefined {
  if (!manifest.bgm) return undefined;
  if (Array.isArray(manifest.bgm)) {
    return manifest.bgm[0] && typeof manifest.bgm[0] === 'object' 
      ? (manifest.bgm[0] as { id?: string }).id 
      : undefined;
  }
  return typeof manifest.bgm === 'string' ? manifest.bgm : undefined;
}

function loadManifest(eraDir: string): EraManifest | null {
  const manifestPath = path.join(eraDir, 'manifest.json');
  if (!fs.existsSync(manifestPath)) return null;
  
  try {
    const content = fs.readFileSync(manifestPath, 'utf-8');
    return JSON.parse(content) as EraManifest;
  } catch {
    return null;
  }
}

function getStatus(manifest: EraManifest | null, format: 'old' | 'new' | 'none'): string {
  if (!manifest) return 'missing';
  if (manifest.status) return manifest.status;
  if (format === 'new') {
    const imgCount = countImages(manifest);
    return imgCount >= 6 ? 'complete' : imgCount > 0 ? 'partial' : 'pending';
  }
  return 'unknown';
}

// ── Main Analysis ─────────────────────────────────────────────────────────────

function analyzeEras(): EraAnalysis[] {
  const results: EraAnalysis[] = [];
  
  if (!fs.existsSync(ERA_ASSETS_DIR)) {
    console.error(`[ERROR] Era assets directory not found: ${ERA_ASSETS_DIR}`);
    return results;
  }
  
  const entries = fs.readdirSync(ERA_ASSETS_DIR, { withFileTypes: true });
  
  for (const entry of entries) {
    if (!entry.isDirectory()) continue;
    if (entry.name === 'index.html') continue;
    
    const eraDir = path.join(ERA_ASSETS_DIR, entry.name);
    const manifestPath = path.join(eraDir, 'manifest.json');
    const format = getManifestFormat(manifestPath);
    const manifest = loadManifest(eraDir);
    const status = getStatus(manifest, format);
    
    // Count actual files
    const files = fs.readdirSync(eraDir)
      .filter(f => f !== 'manifest.json' && !f.endsWith('.json'));
    
    // Determine if complete
    const imageCount = manifest ? countImages(manifest) : 0;
    const hasBgmFlag = hasBgm(manifest);
    const isComplete = imageCount >= 6 && hasBgmFlag;
    const needsAttention = status === 'missing' || status === 'pending' || status === 'partial';
    
    results.push({
      name: entry.name,
      id: manifest?.era_id || manifest?.id || entry.name,
      has_manifest: format !== 'none',
      manifest_format: format,
      status,
      image_count: imageCount,
      has_bgm: hasBgmFlag,
      bgm_name: getBgmName(manifest),
      files,
      is_complete: isComplete,
      needs_attention: needsAttention,
    });
  }
  
  return results.sort((a, b) => a.name.localeCompare(b.name));
}

function generateSummary(eras: EraAnalysis[]): AssetRequirements {
  const complete = eras.filter(e => e.status === 'complete').length;
  const partial = eras.filter(e => e.status === 'partial').length;
  const pending = eras.filter(e => e.status === 'pending').length;
  const missing = eras.filter(e => e.status === 'missing').length;
  
  return {
    era_scenes: {
      total: eras.length,
      complete,
      partial,
      pending,
      missing,
    },
    item_icons: { total: 60, current: 14, gap: 46 },
    building_icons: { total: 21, current: 4, gap: 17 },
    npc_avatars: { total: 34 },
    skill_icons: { total: 28, current: 1, gap: 27 },
    ui_icons: { total: 37, current: 20, gap: 17 },
    bgm: {
      scene: { total: 36, complete: 0, partial: 0, pending: 0, missing: 0 }, // 需要单独统计
      era: { total: 42, complete, partial, pending, missing },
    },
    sfx: { total: 58, current: 15, gap: 43 },
  };
}

function printReport(eras: EraAnalysis[], summary: AssetRequirements): void {
  console.log('╔══════════════════════════════════════════════════════════════════╗');
  console.log('║         素材资源需求分析报告                                      ║');
  console.log('║         Asset Resource Requirements Analysis                       ║');
  console.log('╚══════════════════════════════════════════════════════════════════╝');
  console.log('');
  
  console.log('📊 时代素材总览 (Era Scene Assets Overview)');
  console.log('─'.repeat(70));
  console.log(`  总计子纪元: ${eras.length}`);
  console.log(`  ✅ 已完成 (complete):   ${summary.era_scenes.complete}`);
  console.log(`  ⚠️  部分完成 (partial): ${summary.era_scenes.partial}`);
  console.log(`  ⏳ 待生成 (pending):   ${summary.era_scenes.pending}`);
  console.log(`  ❌ 缺失 (missing):     ${summary.era_scenes.missing}`);
  console.log('');
  
  console.log('📋 各时代素材状态 (By Era)');
  console.log('─'.repeat(70));
  
  for (const era of eras) {
    const statusIcon = era.is_complete ? '✅' : era.needs_attention ? '⚠️' : '❌';
    const formatLabel = era.manifest_format === 'new' ? '[NEW]' : era.manifest_format === 'old' ? '[OLD]' : '[NO-MANIFEST]';
    const imgInfo = `${era.image_count}/6 images`;
    const bgmInfo = era.has_bgm ? '🎵' : '❌';
    
    console.log(`  ${statusIcon} ${formatLabel} ${era.name.padEnd(35)} ${imgInfo.padEnd(12)} ${bgmInfo}`);
  }
  
  console.log('');
  console.log('📦 其他资源需求 (Other Asset Requirements)');
  console.log('─'.repeat(70));
  console.log(`  物品图标: ${summary.item_icons.current}/${summary.item_icons.total} (缺口: ${summary.item_icons.gap})`);
  console.log(`  建筑图标: ${summary.building_icons.current}/${summary.building_icons.total} (缺口: ${summary.building_icons.gap})`);
  console.log(`  NPC头像: ${summary.npc_avatars.total} (fallback)`);
  console.log(`  技能图标: ${summary.skill_icons.current}/${summary.skill_icons.total} (缺口: ${summary.skill_icons.gap})`);
  console.log(`  UI图标: ${summary.ui_icons.current}/${summary.ui_icons.total} (缺口: ${summary.ui_icons.gap})`);
  console.log(`  BGM: 场景${summary.bgm.scene.total}首 + 时代${summary.bgm.era.total}首`);
  console.log(`  SFX: ${summary.sfx.current}/${summary.sfx.total} (缺口: ${summary.sfx.gap})`);
  console.log('');
  
  // Priority items
  const priorityEras = eras.filter(e => e.status === 'pending' || e.status === 'missing').slice(0, 10);
  if (priorityEras.length > 0) {
    console.log('🎯 优先补齐项目 (Priority Items)');
    console.log('─'.repeat(70));
    for (const era of priorityEras) {
      console.log(`  • ${era.name} (${era.status})`);
    }
    console.log('');
  }
}

function saveJsonReport(eras: EraAnalysis[], summary: AssetRequirements): void {
  if (!fs.existsSync(OUTPUT_DIR)) {
    fs.mkdirSync(OUTPUT_DIR, { recursive: true });
  }
  
  const report = {
    generated_at: new Date().toISOString(),
    summary,
    eras,
  };
  
  const outputPath = path.join(OUTPUT_DIR, `asset-requirements-${Date.now()}.json`);
  fs.writeFileSync(outputPath, JSON.stringify(report, null, 2), 'utf-8');
  console.log(`[INFO] JSON report saved: ${outputPath}`);
  
  // Also save latest
  const latestPath = path.join(OUTPUT_DIR, 'latest.json');
  fs.writeFileSync(latestPath, JSON.stringify(report, null, 2), 'utf-8');
  console.log(`[INFO] Latest report saved: ${latestPath}`);
}

// ── CLI Entry Point ────────────────────────────────────────────────────────────

async function main() {
  const args = process.argv.slice(2);
  const outputJson = args.includes('--json');
  const eraFilter = args.find(a => a.startsWith('--era='))?.split('=')[1];
  
  console.log('[INFO] Analyzing era assets...');
  let eras = analyzeEras();
  
  if (eraFilter) {
    eras = eras.filter(e => e.name.includes(eraFilter) || e.id.includes(eraFilter));
    console.log(`[INFO] Filtered to: ${eraFilter}`);
  }
  
  const summary = generateSummary(eras);
  
  if (outputJson) {
    saveJsonReport(eras, summary);
  } else {
    printReport(eras, summary);
  }
}

main().catch(e => {
  console.error('[FATAL]', e);
  process.exit(1);
});
